update simple search

jlmitch5 opened the pull request:
Fixes https://github.com/ansible/awx/issues/5365 https://github.com/ansible/awx/issues/5033 https://github.com/ansible/awx/issues/4785 https://github.com/ansible/awx/issues/4748 https://github.com/ansible/awx/issues/5631

This PR makes some changes to the infrastructure underlying search, implements some new UX based on new components from Patternfly, and updates lists across the app to have the necessary filters.

## FOLLOWUP ISSUES

- https://github.com/ansible/awx/issues/5626
- https://github.com/ansible/awx/issues/5627
- https://github.com/ansible/awx/issues/5628
- https://github.com/ansible/awx/issues/5629

## PF TODOS

There are a few issues and items awaiting implementation on the PF side:
- Support needed for object-based values for checkbox dropdown.  Currently isChecked prop for the individual options will not set unless the value is a string.  All options include both string and value (that django expects to receive) for the options, using both just needs to be wired up in Search.jsx.  Currently an in-progress patternfly-react [issue](https://github.com/patternfly/patternfly-react/issues/3401)
- There are different left/right padding amounts for the new DataToolbar component and the individual list rows.  This is a known issue, and patternfly will be updating CSS so that the padding is consistent.  [issue](https://github.com/patternfly/patternfly-next/issues/2503) currently open to address this in patternfly-core (css)
- Overflow of keys and tag values in the search tags section needs some work.  Currently the values ellipses, but there's no tooltip to show the full valu.  [fixed issue for adding ellipses](https://github.com/patternfly/patternfly-next/issues/2524) in patternfly-next, no issue yet for tooltip.
- There are various small issues being addressed with the visual design of the DataToolbar components in this [PR](https://github.com/patternfly/patternfly-react/issues/3396).  We should be able to take advantage of these fixes without any code changes once they are included in a PF release and we update.

## DONE

- Update COLUMNS to SORT_COLUMNS and SEARCH_COLUMNS
- Update to using new PF Toolbar component (currently an experimental component)
- Change the right-hand input based on the type of key selected on the left-hand side.  In addition to text input, for our MVP we will support:
  - number input
  - select input (multiple-choice configured from UI or Options.  Includes multiple select and boolean-based)
- Update the following lists to have the following keys:

### In this PR:

**Jobs list** (signed off earlier in chat)
  - Name (which is also the name of the job template) - search is ?name=jt
  - Job ID - search is ?id=13
  - Label name - search is ?labels__name=foo
  - Job type (dropdown on right with the different types) ?type = job
  - Created by (username) - search is ?created_by__username=admin
  - Status - search (dropdown on right with different statuses) is ?status=successful

Instances of jobs list include:
  - Jobs list
  - JT completed jobs list

**Organization list**
  - Name - search is ?name=org
  - Created by (username)
  - Modified by (username)

Instances of orgs list include:
  - Orgs list
  - Lookup on Project
  - Lookup on Inventory

**Instance Groups list**
  - Name - search is ?name=ig
  - Credential name - search is ?credentials__name=kubey
  - Created by (username)
  - Modified by (username)

Instance of instance groups list include:
  - Lookup on Org
  - Lookup on JT
  - Lookup on Inventory

**Users list**
  - Username - search is ?username=johndoe
  - First Name - search is ?first_name=John
  - Last Name - search is ?last_name=Doe

Instance of user lists include:
  - User list
  - Org user list
  - Access list (Shares same list in UI.  eventually will be on Org, JT, Project, Credential, Inventory, User, Team)
  - Team add users list
  - Users list in access wizard (to add new roles for a particular list) (Shares same list in UI. eventually will be on Org, JT, Project, Credential)

**Teams list**
  - Name - search is ?name=teamname
  - (only root team list) Org Name - search is ?organization__name=org_of_john_does
  - Created by (username)
  - Modified by (username)

Instance of team lists include:
  - Team list
  - Org team list
  - Assorted teams list in access wizard

**Credentials list**
  - Name 
  - Created by (username)
  - Modified by (username)

Instance of credential lists include:
  - Lookup for JT
  - Lookup for Project

**Projects list**
  - Name - search is ?name=proj
  - Type (dropdown on right with different types) - search is scm_type=git
  - SCM URL - search is ?scm_url=github.com/ansible/test-playbooks
  - Created by (username) - search is ?created_by__username=admin
  - Modified by (username) - search is ?modified_by__username=admin

Instance of project lists include:
  - Project list
  - Lookup for JT

**Templates list**
  - Name - search is ?name=cleanup
  - Type (dropdown on right with different types) - search is ?type=playbook_run
  - Playbook name - search is ?job_template__playbook=debug.yml
  - Launched by (username) - search is ?created_by__username=admin
  - Modified by (username) - search is ?modified_by__username=admin
  - Limit - search is ?job__limit=foo

Instance of template lists include:
  - Template list

**Inventories list**
  - Name - search is ?name=inv
  - Created by (username) - search is ?created_by__username=admin
  - Modified by (username) - search is ?modified_by__username=admin

Instance of inventory lists include:
  - Inventory list
  - Lookup for JT

**Groups list**
  - Name - search is ?name=group_name
  - Created by (username) - search is ?created_by__username=admin
  - Modified by (username) - search is ?modified_by__username=admin
  - Only root groups (boolean choice) - search is ?parents__isnull=true

Instance of group lists include:
  - Group list

**Hosts list**
  - Name - search is ?name=hostname
  - Created by (username) - search is ?created_by__username=admin
  - Modified by (username) - search is ?modified_by__username=admin

Instance of host lists include:
  - Host list

**Notifications list**
  - Name - search is ?name=notification_template_name
  - Type (dropdown on right with different types) - search is ?type=slack
  - Created by (username) - search is ?created_by__username=admin
  - Modified by (username) - search is ?modified_by__username=admin

Instance of notification lists include:
  - Org notification list
  - JT notification list
  - Project notification list

### Not in this PR:

Instances of jobs list include:
  - Host completed jobs list (not implemented yet)
  - Smart Inventory completed jobs list (not implemented yet)
  - JT completed jobs list (not implemented yet)

**Organization list**
  - Team name (of a team in the org) - search is ?teams__name=ansible (doesn't work in api)
  - Username (of a user in the org) - search is ?users__username=johndoe (doesn't work in api)

Instances of orgs list include:
  - User orgs list (not implemented yet)
  - Lookup on Credential (not implemented yet)
  - User access add wizard list (not implemented yet)
  - Team access add wizard list (not implemented yet)

**Instance Groups list**
  - eventual "type" select field (not yet implemented on the model)

Instance of instance groups list include:
  - Instance Groups list (not implemented yet)

**Users list**
  - (if not superfluous, would not include on Team users list) Team Name - search is ?teams__name=team_of_john_does (note API issue: User has no field named "teams")
  - (only for access or permissions list) Role Name - search is ?roles__name=Admin (note API issue: Role has no field "name")
  - (if not superfluous, would not include on Organization users list) Org Name - search is ?organizations__name=org_of_jhn_does (note API Issue: Organization has no field "name")

Instance of user lists include:
  - Team add users list (not yet implemented)

Instance of team lists include:
  - User team list (not yet implemented)

Instance of credential lists include:
  - Credential list (not yet implemented)
  - User access add wizard list (not yet implemented)
  - Team access add wizard list (not yet implemented)

Instance of project lists include:
  - User access add wizard list (not yet implemented)
  - Team access add wizard list (not yet implemented)

Instance of template lists include:
  - Project Templates list (not yet implemented)

Instance of inventory lists include:
  - User access add wizard list (not yet implemented)
  - Team access add wizard list (not yet implemented)

Commit:
update simple search doumentation and plan

wenottingham commented on awx/ui_next/SEARCH.md:
```diff
@@ -24,143 +49,256 @@
+  - Name (which is also the name of the job template) - search is ?name=jt
+  - Job ID - search is ?id=13
+  - Label name - search is ?labels__name=foo
+  - Job type (dropdown on right with the different types) ?type = job
```
is this intended to *not* be a substring search? (`job` is a substring in all job types, which makes it not useful as a distinguisher)

jlmitch5 replied:
@wenottingham this will be a multiple choice selection (not a substring).  the options will be:

Text (string) inputs are substring/case-insensitive (we append __icontains to all those keys).

```
"choices": [
                    [
                        "project_update",
                        "SCM Update"
                    ],
                    [
                        "inventory_update",
                        "Inventory Sync"
                    ],
                    [
                        "job",
                        "Playbook Run"
                    ],
                    [
                        "ad_hoc_command",
                        "Command"
                    ],
                    [
                        "system_job",
                        "Management Job"
                    ],
                    [
                        "workflow_job",
                        "Workflow Job"
                    ]
                ]
```

trahman73 reviewed (commented):
I read through the UX portion of the doc and agree with the implementation. @wenottingham Are you ok with us not being able to repopulate the search string when a search tag is clicked for now?

wenottingham commented:
> I read through the UX portion of the doc and agree with the implementation. @wenottingham Are you ok with us not being able to repopulate the search string when a search tag is clicked for now?

Yes, that's fine for now.

jlmitch5 commented:
Changing to WIP PR, will include updates based on plan

Commit:
update search and sort column configuration

Commit:
remove debugger statement

Commit:
utilize new DataToolbar experimental patternfly components

Commit:
add support for number, boolean, and option-based searches

Commit:
support 1 item in sortColumns

Commit:
update search keys for various lists in the mvp

Commit:
add keys to search on lookups

Commit:
fix issues with unit tests based on search changes

Commit:
swap column types use of shape for exact and fix warnings it found

Commit:
second pass double checking all keys show up in the lists they should

mabashian commented:
It looks like we need to clear out the page number before we do a search.  If you go to page 2 and perform a search that doesn't have 2 pages then an error gets thrown.

![multipage_search](https://user-images.githubusercontent.com/9889020/71849844-05006000-30a1-11ea-9639-46336d551972.gif)

mabashian commented:
@jlmitch5 @trahman73 Not sure if you all talked about this - if you did, feel free to ignore.  I'm wondering if some of these options should be capitalized:

<img width="712" alt="Screen Shot 2020-01-06 at 4 27 10 PM" src="https://user-images.githubusercontent.com/9889020/71850079-7e984e00-30a1-11ea-84ac-e5f81277ff03.png">

`Label name` -> `Label Name`
`Job type` -> `Job Type`
`Created by (username)` -> `Created By (Username)`

Thoughts?

Might want to do the same thing with these options:

<img width="590" alt="Screen Shot 2020-01-06 at 4 37 00 PM" src="https://user-images.githubusercontent.com/9889020/71850663-d2effd80-30a2-11ea-9439-b2b854b3e90f.png">

mabashian commented:
Looks like the jobs list search isn't happy about this field:

<img width="1434" alt="Screen Shot 2020-01-06 at 4 29 48 PM" src="https://user-images.githubusercontent.com/9889020/71850224-d2a33280-30a1-11ea-96ee-27e1dd094a96.png">

mabashian commented:
I might have broken this when I upgraded pf:

![filter_job_type](https://user-images.githubusercontent.com/9889020/71850427-447b7c00-30a2-11ea-84a6-a483a3b833c0.gif)

Are those checkboxes supposed to get checked?

mabashian commented:
This is probably a PF thing - it seems like long key names get chopped off in the search tag:

<img width="941" alt="Screen Shot 2020-01-06 at 4 34 21 PM" src="https://user-images.githubusercontent.com/9889020/71850526-7856a180-30a2-11ea-8021-883e37677fb6.png">

jlmitch5 commented:
> This is probably a PF thing - it seems like long key names get chopped off in the search tag:
> 
> <img alt="Screen Shot 2020-01-06 at 4 34 21 PM" width="941" src="https://user-images.githubusercontent.com/9889020/71850526-7856a180-30a2-11ea-8021-883e37677fb6.png">

yep!  there looking at fixing this one -- https://github.com/patternfly/patternfly-next/issues/2524

> I might have broken this when I upgraded pf:
> 
> ![filter_job_type](https://user-images.githubusercontent.com/9889020/71850427-447b7c00-30a2-11ea-84a6-a483a3b833c0.gif)
> 
> Are those checkboxes supposed to get checked?

ugh yeah, this one's come back a couple times, let me see if I can figure it out



> Looks like the jobs list search isn't happy about this field:

whoops!  I meant `labels__name`, not `label__name`

Commit:
update search based on pr feedback:
- fix InventoryHost list inadverdent breakage due to merge conflict
- update label__name search key to labels__name
- always snap to page 1 when a search happens
- udpate capitalization of keys to be consistent
- remove isDefault from sort col items in tests

mabashian commented on awx/ui_next/src/util/qs.js:
```diff
@@ -193,6 +193,7 @@ function removeParam(oldVal, deleteVal) {
  * @return {object} merged namespaced params object
  */
 export function mergeParams(oldParams, newParams) {
+  debugger;
```
Looks like you left a `debugger;` in here

mabashian commented:
On the instance groups list - it looks like that endpoint doesn't have `created_by` or `modified_by` fields:

<img width="1589" alt="Screen Shot 2020-01-08 at 10 33 38 AM" src="https://user-images.githubusercontent.com/9889020/71991021-73106880-3202-11ea-96d7-56d57e4beefb.png">

<img width="1680" alt="Screen Shot 2020-01-08 at 10 34 42 AM" src="https://user-images.githubusercontent.com/9889020/71991045-815e8480-3202-11ea-8340-66c0eed4652c.png">

Not sure if we need to interface with the api team to get these fields added or?

mabashian commented:
Not sure if this is related to my previous comment about the checkboxes in the dropdown but the project -> type dropdown has a blank field in it:

<img width="1070" alt="Screen Shot 2020-01-08 at 10 43 57 AM" src="https://user-images.githubusercontent.com/9889020/71992751-e9619a80-3203-11ea-811d-9dd173921a80.png">

mabashian commented:
I think this is benign but there's an error thrown to the console when you first load the projects list:

<img width="1676" alt="Screen Shot 2020-01-08 at 10 46 01 AM" src="https://user-images.githubusercontent.com/9889020/71992901-2f1e6300-3204-11ea-92ee-0d77faa95e1e.png">

mabashian commented:
Looks like the `Clear all filters` doesn't get hidden after removing the last project type tag:

![project_type](https://user-images.githubusercontent.com/9889020/71993147-9a683500-3204-11ea-9721-4c0fa25036da.gif)

mabashian commented:
<img width="863" alt="Screen Shot 2020-01-08 at 10 59 32 AM" src="https://user-images.githubusercontent.com/9889020/71994182-4d855e00-3206-11ea-8225-313500eead73.png">

Wondering if we should uppercase/convert these to `Job Template`/`Workflow Job Template`?


**Edit** 

I see where you have titles for each of those options with nice pretty text - disregard this comment.  I think this is related to the previous comment about those dropdown options and the checkboxes.

mabashian commented on awx/ui_next/src/components/AddRole/SelectResourceStep.test.jsx:
```diff
@@ -6,20 +6,31 @@ import { sleep } from '../../../testUtils/testUtils';
-    { name: 'Username', key: 'username', isSortable: true, isSearchable: true },
+  const searchColumns = [
+    {
+      name: 'Username',
```
Mark for translation

mabashian commented on awx/ui_next/src/components/AddRole/SelectResourceStep.test.jsx:
```diff
@@ -6,20 +6,31 @@ import { sleep } from '../../../testUtils/testUtils';
+
+  const sortColumns = [
+    {
+      name: 'Username',
```
Mark for translation

keithjgrant commented:
I think it might look a little nicer if we can indent the active search tags to the right a little (i.e. leave a gap where the checkboxes are in other rows)

![Screen Shot 2020-01-08 at 4 01 04 PM](https://user-images.githubusercontent.com/410794/72026423-33607580-3230-11ea-8505-b03465a0b6a5.png)

keithjgrant commented on awx/ui_next/src/components/Sort/Sort.jsx:
```diff
@@ -17,59 +19,47 @@ import {
+      sortOrder = 'ascending';
+    }
+
+    if (qsConfig.integerFields.filter(field => field === sortKey).length) {
```
We're doing this logic in a few places. It's probably worth adding a helper function to qs (something like `isFieldNumeric(qsConfig, fieldName)`). It's a little cleaner to use `Array.find()`instead of `.filter()`, too.

It might be worth doing the same with the `sortOrder` check above.

jlmitch5 replied:
I can update to find, but I don't think I agree that moving this up a level is an improvement.

`qsConfig.integerFields.find(field => field === sortKey)` vs `qsConfig.isFieldNumeric(qsConfig, sortKey)`

This concept is only used in two places (search and sort components).  I don't find one cleaner than the other, and I don't think making the qs API more complex for this case is worth it.

keithjgrant commented on awx/ui_next/src/util/qs.js:
```diff
@@ -14,6 +14,10 @@ export function getQSConfig(
     throw new Error('a QS namespace is required');
   }
+  // if order_by isn't passed, default to name
+  if (!Object.keys(defaultParams).filter(key => key === 'order_by').length) {
```
Is there a reason you don't just check `if (!defaultParams.order_by)`?

jlmitch5 replied:
Good catch, much more direct way to get what I'm looking for hah

keithjgrant reviewed (commented):
This is looking good. Nice and clean, too. I kind of wonder if we could move all the column definitions out of components and into a common config file somehow, but I can't think of an easy way to i18n everything if we do.

jlmitch5 commented:
@keithjgrant 

> I think it might look a little nicer if we can indent the active search tags to the right a little (i.e. leave a gap where the checkboxes are in other rows)

There's an open pf design issue to bring back result count, which will adjust things in a way that feels better.  https://github.com/patternfly/patternfly-design/issues/780

> I kind of wonder if we could move all the column definitions out of components and into a common config file somehow, but I can't think of an easy way to i18n everything if we do.

I'm not sure I agree this would be preferred.  As it currently stands, our list components configure things like what items show up in the data rows, and I believe what you can search/sort on for a particular list to be a similar concern.

Commit:
more ui_next search pr feedback:
- updae .filter().length calls to .find()
- fix ProjectList errors

Commit:
fix issue with checkbox state of select based search key

Commit:
remove created by/modified by search keys from ig lookup

Commit:
add limit search key to jobs list

Commit:
fix prettier

Commit:
update sort iconography
